SPM: update DML SQL Bind and baseline capture description (#5088)
pingcap/docs: sql-plan-management.md

{{< copyable "sql" >}}

```sql
CREATE [GLOBAL | SESSION] BINDING FOR BindableStmt USING BindableStmt
```

This statement binds SQL execution plans at the GLOBAL or SESSION level. Currently, supported bindable SQL statements (BindableStmt) in TiDB include `SELECT`, `DELETE`, `UPDATE`, and `INSERT` / `REPLACE` with `SELECT` subqueries.

Specifically, two types of these statements cannot be bound to execution plans due to syntax conflicts. See the following examples:

```sql

> -- Type one: Statements that use the `join` keyword to get the Cartesian product but not specify the associated columns with the `using` keyword.

Reviewer: ```suggestion
-- Type one: Statements that use the `join` keyword but not specify the associated columns with the `using` keyword, to get the Cartesian product.
```

Author: ```suggestion
-- Type one: Statements that use the `join` keyword but do not specify the associated columns with the `using` keyword to get the Cartesian product.
```
How about this?

Reviewer: 使用 join 但不没有 using，才会得到笛卡尔积，而不是说使用 using 关键字能得到笛卡尔积。

create global binding for
    select * from t t1 join t t2
using
    select * from t t1 join t t2;
-- Type two: `DELETE` statements that contain the `using` keyword.
create global binding for
    delete from t1 using t1 join t2 on t1.a = t2.a
using
    delete from t1 using t1 join t2 on t1.a = t2.a;
```

You can bypass syntax conflicts by using equivalent statements. For example, you can rewrite the above statements in the following ways:

```sql
-- First rewrite of type one statements: Add a `using` clause for the `join` keyword.
create global binding for
    select * from t t1 join t t2 using (a)
using
    select * from t t1 join t t2 using (a);

> -- Second rewrite of type one statements: Delete the `join` keyword.

Author: ```suggestion

-- Second rewrite of type one statements: Delete the `join` keyword.
```

create global binding for
    select * from t t1, t t2
using
    select * from t t1, t t2;

> -- Rewrite of type two statements: Remove the `using` keyword from the `delete` statement.

Author: ```suggestion

-- Rewrite of type two statements: Remove the `using` keyword from the `delete` statement.
```

create global binding for
    delete t1 from t1 join t2 on t1.a = t2.a
using
    delete t1 from t1 join t2 on t1.a = t2.a;
```

> **Note:**
> When creating execution plan bindings for `INSERT` / `REPLACE` statements with `SELECT` subqueries, you need to specify the optimizer hints you want to bind in the `SELECT` subquery, not after the `INSERT` / `REPLACE` keyword. Otherwise, the optimizer hints do not take effect as intended.

Here are two examples:

```sql
-- The hint takes effect in the following statement.
create global binding for
    insert into t1 select * from t2 where a > 1 and b = 1
using
    insert into t1 select /*+ use_index(@sel_1 t2, a) */ * from t2 where a > 1 and b = 1;

> -- The hint cannot take effect in the following statement.

Author: ```suggestion

-- The hint cannot take effect in the following statement.
```

create global binding for
    insert into t1 select * from t2 where a > 1 and b = 1
using
    insert /*+ use_index(@sel_1 t2, a) */ into t1 select * from t2 where a > 1 and b = 1;
```

> If you do not specify the scope when creating an execution plan binding, the default scope is SESSION. The TiDB optimizer standardizes bound SQL statements and stores them in the system table. When processing SQL queries, if a standardized statement matches one of the bound SQL statements in the system table and the system variable `tidb_use_plan_baselines` is set to `on` (the default value is `on`), TiDB then uses the corresponding optimizer hint for this statement. If there are multiple matchable execution plans, the optimizer choose the least costly one to bind.

Reviewer: ```suggestion
If you do not specify the scope when creating an execution plan binding, the default scope is SESSION. The TiDB optimizer normalizes bound SQL statements and stores them in the system table. When processing SQL queries, if a normalized statement matches one of the bound SQL statements in the system table and the system variable `tidb_use_plan_baselines` is set to `on` (the default value is `on`), TiDB then uses the corresponding optimizer hint for this statement. If there are multiple matchable execution plans, the optimizer choose the least costly one to bind.
```

> `Standardization` is a process that converts a constant in an SQL statement to a variable parameter and explicitly specifies the database for tables referenced in the query, with standardized processing on the spaces and line breaks in the SQL statement. See the following example:

Reviewer: ```suggestion
`Normalization` is a process that converts a constant in an SQL statement to a variable parameter and explicitly specifies the database for tables referenced in the query, with standardized processing on the spaces and line breaks in the SQL statement. See the following example:
```

```sql
select * from t where a >    1

> -- Standardized:

Reviewer: ```suggestion
-- Normalized:
```

When the first `select` statement is being executed, the optimizer adds the `sm_join(t1, t2)` hint to the statement through the binding in the GLOBAL scope. The top node of the execution plan in the `explain` result is MergeJoin. When the second `select` statement is being executed, the optimizer uses the binding in the SESSION scope instead of the binding in the GLOBAL scope and adds the `hash_join(t1, t2)` hint to the statement. The top node of the execution plan in the `explain` result is HashJoin.

Each standardized SQL statement can have only one binding created using `CREATE BINDING` at a time. When multiple bindings are created for the same standardized SQL statement, the last created binding is retained, and all previous bindings (created and evolved) are marked as deleted. But session bindings and global bindings can coexist and are not affected by this logic.

In addition, when you create a binding, TiDB requires that the session is in a database context, which means that a database is specified when the client is connected or `use ${database}` is executed.

> > **Note:**

Author: It is suggested to be marked as a note in https://github.com/pingcap/docs/pull/1249#discussion_r296576774, but I agreed with you. I will update it later.

>

> > The text must be the same before and after normalization and hint removal for both the original SQL statement and the bound statement, or the binding will fail. Take the following examples:

Reviewer: Yep.

> > The text must be the same before and after standardization and hint removal for both the original SQL statement and the bound statement, or the binding will fail. Take the following examples:

Reviewer: ```suggestion
> The text must be the same before and after normalization and hint removal for both the original SQL statement and the bound statement, or the binding will fail. Take the following examples:
```

>
> Because the automatic binding creation function relies on [Statement Summary](/statement-summary-tables.md), make sure to enable Statement Summary before using automatic binding.

> After automatic binding creation is enabled, the historical SQL statements in the Statement Summary are traversed every `bind-info-lease` (the default value is `3s`), and baseline capturing is automatically created for SQL statements that appear at least twice. For these SQL statements, TiDB automatically binds the execution plan recorded in Statement Summary.

Reviewer: ```suggestion
After automatic binding creation is enabled, the historical SQL statements in the Statement Summary are traversed every `bind-info-lease` (the default value is `3s`), and a binding is automatically created for SQL statements that appear at least twice. For these SQL statements, TiDB automatically binds the execution plan recorded in Statement Summary.
```

> However, TiDB does not automatically capture bindings from the baseline for the following types of SQL statements:

Reviewer: ```suggestion
However, TiDB does not automatically capture bindings for the following types of SQL statements:
```

- `EXPLAIN` and `EXPLAIN ANALYZE` statements.
- SQL statements executed internally in TiDB, such as `SELECT` queries used for automatically loading statistical information.
- SQL statements that are bound to a manually created execution plan.

> For `PREPARE` / `EXECUTE` statements and queries executed with binary protocols, TiDB automatically captures bindings for the real query statements, not for the `PREPARE` / `EXECUTE` statements.

Author: ```suggestion
For `PREPARE` / `EXECUTE` statements and for queries executed with binary protocols, TiDB automatically captures bindings for the real query statements, not for the `PREPARE` / `EXECUTE` statements.
```

> For `PREPARE` / `EXECUTE` statements and queries executed with binary protocols, TiDB automatically captures bindings from the baseline for real query statements, not for the `PREPARE` / `EXECUTE` statements.

Reviewer: ```suggestion
For `PREPARE` / `EXECUTE` statements and queries executed with binary protocols, TiDB automatically captures bindings for the real query statements, not for the `PREPARE` / `EXECUTE` statements.
```
